In [8]:
import torch
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import matplotlib as plt
import seaborn as sns
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

In [20]:
print("PyTorch version:", torch.__version__)
print("CUDA available:", torch.cuda.is_available())
print("GPU Name:", torch.cuda.get_device_name(0) if torch.cuda.is_available() else "No GPU found")

PyTorch version: 2.5.1+cu121
CUDA available: True
GPU Name: NVIDIA GeForce RTX 4060 Laptop GPU


In [10]:
df = pd.read_csv("creditcard.csv")
df

Time         V1         V2  ...       V28  Amount  Class
0            0.0  -1.359807  -0.072781  ... -0.021053  149.62      0
1            0.0   1.191857   0.266151  ...  0.014724    2.69      0
2            1.0  -1.358354  -1.340163  ... -0.059752  378.66      0
3            1.0  -0.966272  -0.185226  ...  0.061458  123.50      0
4            2.0  -1.158233   0.877737  ...  0.215153   69.99      0
...          ...        ...        ...  ...       ...     ...    ...
284802  172786.0 -11.881118  10.071785  ...  0.823731    0.77      0
284803  172787.0  -0.732789  -0.055080  ... -0.053527   24.79      0
284804  172788.0   1.919565  -0.301254  ... -0.026561   67.88      0
284805  172788.0  -0.240440   0.530483  ...  0.104533   10.00      0
284806  172792.0  -0.533413  -0.189733  ...  0.013649  217.00      0

[284807 rows x 31 columns]

Here, time is the number of seconds elapsed a particular transaction from the first transaction.

All the "V" features are anonymized with PCA to make sure no banking details of the customers get leaked. Those features are previously the banking details.

PCA -> Principal components

It’s like scrambling the columns, but in a very smart way that retains the structure, patterns, and most meaningful variation.

Note : The new axes are uncorrelated.

Here, the amount column is transaction amount

Class :- 1 for fraudulent, 0 otherwise.

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [17]:
df['Class'].value_counts(normalize=False)

Class
0    284315
1       492
Name: count, dtype: int64

There are only 492 fraudulent transactions in 284807 total transcations.

In [18]:
percentage = (492/283807) * 100
percentage

0.17335724629766003

>So, roughly 0.173 percentage of the overall transcations are fraudulent.

>That means, the model is gonna give 99.83 % accuracy regardless but that won't be useful as it's gonna miss the fraudluent transactions.

>The aim of this project won't be to acquire a better accuracy but it will rely more on precision, recall and, F1 score.

### Working with data

In [22]:
X = df.iloc[ :, 0:-1].values
y = df.iloc[ :, -1].values